src/containers/cyclicbuffer.d

```diff
@@ -0,0 +1,828 @@
+ *
+ * Params:
+ *     T = the array element type
+ *     Allocator = the allocator to use. Defaults to `Mallocator`.
+ *     supportGC = true if the container should support holding references to GC-allocated memory.
+ */
+struct CyclicBuffer(T, Allocator = Mallocator, bool supportGC = shouldAddGCRange!T)
+{
+	@disable this(this);
+
+	private import std.conv : emplace;
+	private import std.experimental.allocator.common : stateSize;
+	private import std.traits : isImplicitlyConvertible, hasElaborateDestructor;
+
+	static if (stateSize!Allocator != 0)
+	{
+		/// No default construction if an allocator must be provided.
+		@disable this();
+
+		/**
+		 * Use the given `allocator` for allocations.
+		 */
+		this(Allocator allocator) nothrow pure @safe @nogc
+		in
+		{
+			assert(allocator !is null, "Allocator must not be null");
+		}
+		body
+		{
+			this.allocator = allocator;
+		}
+	}
+
+	~this()
+	{
+		clear();
+		static if (useGC)
+		{
+			import core.memory : GC;
+			GC.removeRange(storage.ptr);
+		}
+		allocator.deallocate(storage);
+	}
+
+	/**
+	 * Removes all contents from the buffer.
+	 */
+	void clear()
+	{
+		if (!empty)
+		{
+			static if (hasElaborateDestructor!T)
+			{
+				if (start <= end)
+					foreach (ref item; storage[start .. end + 1])
+						.destroy(item);
+				else
+				{
+					foreach (ref item; storage[start .. $])
+						.destroy(item);
+					foreach (ref item; storage[0 .. end + 1])
+						.destroy(item);
+				}
+			}
+			start = (end + 1) % capacity;
+			_length = 0;
+		}
+	}
+
+	/**
+	 * Ensures capacity is at least as large as specified.
+	 */
+	size_t reserve(size_t newCapacity)
+	{
+		immutable oldCapacity = capacity;
+		if (newCapacity <= oldCapacity)
+			return oldCapacity;
+		auto old = storage;
+		if (oldCapacity == 0)
+			storage = cast(typeof(storage)) allocator.allocate(newCapacity * T.sizeof);
+		else
+		{
+			auto a = cast(void[]) old;
+			allocator.reallocate(a, newCapacity * T.sizeof);
+			storage = cast(typeof(storage)) a;
+		}
+		static if (useGC)
+		{
+			import core.memory : GC;
+			//Add, then remove. Exactly in that order.
+			GC.addRange(storage.ptr, newCapacity * T.sizeof);
+			GC.removeRange(old.ptr);
+		}
+		if (empty)
+			end = (start - 1 + capacity) % capacity;
+		else if (start > end)
+		{
+			import std.algorithm.mutation : moveEmplaceAll;
+			immutable prefix = end + 1;
+			immutable suffix = oldCapacity - start;
+			if (prefix <= suffix)
+			{
+				immutable space = newCapacity - oldCapacity;
+				if (space >= prefix)
+				{
+					moveEmplaceAll(storage[0 .. prefix], storage[oldCapacity .. $]);
+					end += oldCapacity;
+				}
+				else
+				{
+					moveEmplaceAll(storage[0 .. space], storage[oldCapacity .. $]);
+					end -= space;
+					moveEmplaceAll(storage[space .. prefix], storage[0 .. end + 1]);
+				}
+			}
+			else
+			{
+				import std.range : retro;
+				moveEmplaceAll(
+					retro(storage[start .. oldCapacity]),
+					retro(storage[$ - suffix .. $]),
```

Comment:
The tests do not break when I change this else block to
```d
else
{
	moveEmplaceAll(
		storage[start .. oldCapacity],
		storage[$ - suffix .. $],
	);
	start = newCapacity - suffix;
}
```
What is the purpose of running `retro` on both ranges?